docs/proposals/manifest-hydrator.md

```diff
@@ -0,0 +1,364 @@
+      path: environments/test/west
+    writeTo:
+      targetBranch: environments/test
+      path: west
+---
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: test-east
+spec:
+  sourceHydrator:
+    drySource:
+      repoURL: https://github.com/argoproj/argocd-example-apps
+      targetRevision: main
+      path: environments/test/east
+    writeTo:
+      targetBranch: environments/prod
+      path: east
+---
+### PROD APPS ###
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: prod-west
+spec:
+  sourceHydrator:
+    drySource:
+      repoURL: https://github.com/argoproj/argocd-example-apps
+      targetRevision: main
+      path: environments/prod/west
+    writeTo:
+      targetBranch: environments/prod
+      path: west
+---
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: prod-east
+spec:
+  sourceHydrator:
+    drySource:
+      repoURL: https://github.com/argoproj/argocd-example-apps
+      targetRevision: main
+      path: environments/prod/east
+    writeTo:
+      targetBranch: environments/prod
+      path: east
+---
+```
+
+Each commit to the dry branch will result in a commit to up to three branches. Each commit to an environment branch will contain changes for west, east, or both (depending on which is affected). Changes originating from a single dry commit are always grouped into a single hydrated commit.
+
+### Handling External Values Files
+
+Since only one source may be used in as the dry source, the multi-source approach to external Helm values files will not work here. Instead, we'll recommend that users use the umbrella chart approach. The main reasons for multi-source as an alternative were convenience (no need to maintain the parent chart) and resolving issues with authentication to dependency charts. We believe the simplification is worth the cost of convenience, and we can address the auth issues as standalone bugs.
+
+An earlier iteration of this proposal attempted to preserve the multi-source style of external value file inclusion by introducing a "magic" `.argocd-hydrator.yaml` file containing `additionalSources` to reference the Helm chart. In the end, it felt like we were re-implementing Helm's dependencies feature or git submodules. It's better to just rely on one of those existing tools.
+
+### `.argocd-source.yaml` Support
+
+The `spec.sourceHydrator.drySource` field contains only three fields: `repoURL`, `targetRevision`, and `path`.
+
+`spec.source` contains a number of fields for configuring manifest hydration tools (`helm`, `kustomize`, and `directory`). That functionality is still available for `spec.sourceHydrator`. But instead of being configured in the Application CR, those values are set in `.argocd-source.yaml`, an existing "override" mechanism for `spec.source`. By requiring that this configuration be set in `.argocd-source.yaml`, we respect the principle that all changes must be made in git instead of in the Application CR.
```

Comment:
I think we should default to the Application Source configuration if the `.argocd-source.yaml` file is not specified. It is possible to define Application CR directly in Git already. I think multiple users of ArgoCD use ApplicationSet/Application defined in Helm charts, but ultimately git. Making the source configuration forcibly in git could force people to introduce additional configuration tools to manage source configs as git files instead of using their current mechanism.

Reply:
I think the implementation should be very opinionated that "every change in hydrated manifests originates from some change in a single DRY source." All the benefits of GitOps arise from the assumption that the entire history of the developer's intended system state is represented by a series of commits on a git repo. In my opinion, the failings of GitOps arise from implementations which cause that assumption to be untrue. 

The hydrated manifest history should be an expression of the developer's intent recorded as the history of a single source of truth. Any external augmentation of the developer's intent should be implemented as a mutating webhook , a controller, or some tool which makes changes to the DRY source (e.g. image updater with pushback enabled). Injecting information into the hydrated manifests from the Application CR or from an ApplicationSet (even if its source of truth is _some_ git source) muddies the history of the hydrated manifests by injecting information that is not the developer's intent.

Injection of external information also makes diff previewing less reliable. If I merge a PR based on inspecting the preview diff, and some external system comes and overrides some parameter, I can no longer be confident that what I previewed is what will actually be deployed.

tl;dr: there's a lot of value in being opinionated, but it will alienate some use cases. The feature is opt-in and available for users who want the full benefits of GitOps, but the old way is available for those who need a less opinionated solution and can live with the downsides.

Reply:
Added an "Opinions" section to detail the justification.